server/src/test/java/org/opensearch/action/search/AbstractSearchAsyncActionTests.java

```diff
@@ -328,6 +381,102 @@ private static ArraySearchPhaseResults<SearchPhaseResult> phaseResults(
         return phaseResults;
     }
 
+    public void testOnShardFailurePhaseDoneFailure() throws InterruptedException {
+        final Index index = new Index("test", UUID.randomUUID().toString());
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicBoolean fail = new AtomicBoolean(true);
+
+        final SearchShardIterator[] shards = IntStream.range(0, 5 + randomInt(10))
+            .mapToObj(i -> new SearchShardIterator(null, new ShardId(index, i), List.of("n1", "n2", "n3"), null, null, null))
+            .toArray(SearchShardIterator[]::new);
+
+        SearchRequest searchRequest = new SearchRequest().allowPartialSearchResults(true);
+        searchRequest.setMaxConcurrentShardRequests(1);
+
+        final ArraySearchPhaseResults<SearchPhaseResult> queryResult = new ArraySearchPhaseResults<>(shards.length);
+        AbstractSearchAsyncAction<SearchPhaseResult> action = createAction(
+            searchRequest,
+            queryResult,
+            new ActionListener<SearchResponse>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (fail.compareAndExchange(true, false)) {
+                        try {
+                            throw new NullPointerException("Simulated exception");
+                        } finally {
+                            executor.submit(() -> latch.countDown());
```

Comment:
Why does this countDown() call have to happen on a separate thread? I think this is the only reason the executor is wired in here and it's not obvious to me why you can't just synchronously invoke `countDown()` here.

Reply:
This is a good one, the execution of this callback is happening inside `executor` thread, so for predictability, we have to make sure the current thread finishes and we could release the latch. Otherwise, the current thread will unblock the latch, but the processing may not be finished just yet, making tests flaky.

Reply:
Ah, I missed that the executor was being used to fork off the callbacks as well.